query/workmanager.go

```diff
@@ -0,0 +1,452 @@
+	// the workDispatcher.
+	newBatches chan *batch
+
+	// jobResults is the common channel where results from queries from all
+	// workers will be sent.
+	jobResults chan *jobResult
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// Compile time check to ensure WorkManager satisfies the Dispatcher interface.
+var _ Dispatcher = (*WorkManager)(nil)
+
+// New returns a new WorkManager with the regular worker implementation.
+func New(cfg *Config) *WorkManager {
+	return &WorkManager{
+		cfg:        cfg,
+		newBatches: make(chan *batch),
+		jobResults: make(chan *jobResult),
+		quit:       make(chan struct{}),
+	}
+}
+
+// Start starts the WorkManager.
+func (w *WorkManager) Start() error {
+	w.wg.Add(1)
+	go w.workDispatcher()
+
+	return nil
+}
+
+// Stop stops the WorkManager and all underlying goroutines.
+func (w *WorkManager) Stop() error {
+	close(w.quit)
+	w.wg.Wait()
+
+	return nil
+}
+
+// workDispatcher receives batches of queries to be performed from external
+// callers, and dispatches these to active workers.  It makes sure to
+// prioritize the queries in the order they come in, such that early queries
+// will be attemped completed first.
+//
+// NOTE: MUST be run as a goroutine.
+func (w *WorkManager) workDispatcher() {
+	defer w.wg.Done()
+
+	// Get a peer subscription. We do it in this goroutine rather than
+	// Start to avoid a deadlock when starting the WorkManager fetches the
+	// peers from the server.
+	peersConnected, cancel, err := w.cfg.ConnectedPeers()
+	if err != nil {
+		log.Errorf("Unable to get connected peers: %v", err)
+		return
+	}
+	defer cancel()
+
+	// Init a work queue which will be used to sort the incoming queries in
+	// a first come first served fashion. We use a heap structure such
+	// that we can efficiently put failed queries back in the queue.
+	work := &workQueue{}
+	heap.Init(work)
+
+	type batchProgress struct {
+		timeout <-chan time.Time
+		rem     int
+		errChan chan error
+	}
+
+	// We set up a batch index counter to keep track of batches that still
+	// have queries in flight. This lets us track when all queries for a
+	// batch have been finished, and return an (non-)error to the caller.
+	batchIndex := uint64(0)
+	currentBatches := make(map[uint64]*batchProgress)
+
+	// When the work dispatcher exits, we'll loop through the remaining
+	// batches and send on their error channel.
+	defer func() {
+		for _, b := range currentBatches {
+			b.errChan <- ErrWorkManagerShuttingDown
+		}
+	}()
+
+	// We set up a counter that we'll increase with each incoming query,
+	// and will serve as the priority of each. In addition we map each
+	// query to the batch they are part of.
+	queryIndex := uint64(0)
+	currentQueries := make(map[uint64]uint64)
+
+	workers := make(map[string]*activeWorker)
+
+Loop:
+	for {
+		// If the work queue is non-empty, we'll take out the first
+		// element in order to distribute it to a worker.
+		if work.Len() > 0 {
```

Comment:
One thing about this loop, is that when the job queue is non-empty, and all peers are handling a job, then it'll spin rather rapidly. I don't think this is blocking, but it can be resolved by waiting on a sort of semaphore channel `make(chan struct{}, len(peersConnected))`. This channel would be populated with a `struct{}{}` for each peer to start with, each time we hand a job over we remove an element, and each time we get a response we re-add an element. 

The end result is that we can `select` at the top using this channel, avoiding a tight spin loop when all the peers are saturated. So something like: 
```go
if work.Len() > 0 {
     // Wait for an open slot to become available. 
    select {
        // Free lost, can hand out work. 
        case <-workSema:
        
        case <-quit:
    }
}
```

We re-add to this channel each time we receive a response. Each time the peer set changes, we re-size the channel (re-create) to match the total number of peers.

Reply:
this shouldn't be the case, as `freeWorkers` will be empty, and we'll immediately move down to the below select waiting for a result or a new peer.

Agreed that this could be made more clear, but it will change a bit in the follow up adding more slots to each worker, so will revisit then.